### PR TITLE
Fix out-of-band deserialization of actor handle

### LINKED ISSUE #22780: (Bad error msg) Ref count check failure when passing actor handle serialized
The following script check fails on master. Most likely this is since Tune is pickling the actor handle. We should handle this better / raise a better error message.

```
import ray
from ray import tune


@ray.remote
class Actor:
    pass

a = Actor.remote()

def f(config):
    pass

tune.run(
    f,
    config={
        "actor": a,
    },
)
```

```
(f pid=1393046) *** SIGSEGV received at time=1646251166 on cpu 7 ***
(f pid=1393046) PC: @     0x7fdf19cc9c88  (unknown)  absl::lts_20211102::container_internal::raw_hash_set<>::find<>()
(f pid=1393046)     @     0x7fdf1b3f2210  2140023952  (unknown)
(f pid=1393046)     @     0x7fdf19cd05bd        128  ray::core::ReferenceCounter::GetAndClearLocalBorrowersInternal()
(f pid=1393046)     @     0x7fdf19cd07c4        128  ray::core::ReferenceCounter::GetAndClearLocalBorrowersInternal()
(f pid=1393046)     @     0x7fdf19cd07c4        128  ray::core::ReferenceCounter::GetAndClearLocalBorrowersInternal()
(f pid=1393046)     @     0x7fdf19cd07c4        128  ray::core::ReferenceCounter::GetAndClearLocalBorrowersInternal()
(f pid=1393046)     @     0x7fdf19cd07c4        128  ray::core::ReferenceCounter::GetAndClearLocalBorrowersInternal()
(f pid=1393046)     @     0x7fdf19cd07c4        128  ray::core::ReferenceCounter::GetAndClearLocalBorrowersInternal()
(f pid=1393046)     @     0x7fdf19cd07c4        128  ray::core::ReferenceCounter::GetAndClearLocalBorrowersInternal()
(f pid=1393046)     @     0x7fdf19cd07c4        128  ray::core::ReferenceCounter::GetAndClearLocalBorrowersInternal()
(f pid=1393046)     @     0x7fdf19cd07c4        128  ray::core::ReferenceCounter::GetAndClearLocalBorrowersInternal()
(f pid=1393046)     @     0x7fdf19cd07c4        128  ray::core::ReferenceCounter::GetAndClearLocalBorrowersInternal()
(f pid=1393046)     @     0x7fdf19cd07c4        128  ray::core::ReferenceCounter::GetAndClearLocalBorrowersInternal()
(f pid=1393046)     @     0x7fdf19cd07c4        128  ray::core::ReferenceCounter::GetAndClearLocalBorrowersInternal()
(f pid=1393046)     @     0x7fdf19cd07c4        128  ray::core::ReferenceCounter::GetAndClearLocalBorrowersInternal()
(f pid=1393046)     @     0x7fdf19cd07c4        128  ray::core::ReferenceCounter::GetAndClearLocalBorrowersInternal()
(f pid=1393046)     @     0x7fdf19cd07c4        128  ray::core::ReferenceCounter::GetAndClearLocalBorrowersInternal()
(f pid=1393046)     @     0x7fdf19cd07c4        128  ray::core::ReferenceCounter::GetAndClearLocalBorrowersInternal()
(f pid=1393046)     @     0x7fdf19cd07c4        128  ray::core::ReferenceCounter::GetAndClearLocalBorrowersInternal()
(f pid=1393046)     @     0x7fdf19cd07c4        128  ray::core::ReferenceCounter::GetAndClearLocalBorrowersInternal()
(f pid=1393046)     @     0x7fdf19cd07c4        128  ray::core::ReferenceCounter::GetAndClearLocalBorrowersInternal()
(f pid=1393046)     @     0x7fdf19cd07c4        128  ray::core::ReferenceCounter::GetAndClearLocalBorrowersInternal()
(f pid=1393046)     @     0x7fdf19cd07c4        128  ray::core::ReferenceCounter::GetAndClearLocalBorrowersInternal()
(f pid=1393046)     @     0x7fdf19cd07c4        128  ray::core::ReferenceCounter::GetAndClearLocalBorrowersInternal()
(f pid=1393046)     @     0x7fdf19cd07c4        128  ray::core::ReferenceCounter::GetAndClearLocalBorrowersInternal()
(f pid=1393046)     @     0x7fdf19cd07c4        128  ray::core::ReferenceCounter::GetAndClearLocalBorrowersInternal()
(f pid=1393046)     @     0x7fdf19cd07c4        128  ray::core::ReferenceCounter::GetAndClearLocalBorrowersInternal()
(f pid=1393046)     @     0x7fdf19cd07c4        128  ray::core::ReferenceCounter::GetAndClearLocalBorrowersInternal()
(f pid=1393046)     @     0x7fdf19cd07c4        128  ray::core::ReferenceCounter::GetAndClearLocalBorrowersInternal()
(f pid=1393046)     @     0x7fdf19cd07c4        128  ray::core::ReferenceCounter::GetAndClearLocalBorrowersInternal()
(f pid=1393046)     @     0x7fdf19cd07c4        128  ray::core::ReferenceCounter::GetAndClearLocalBorrowersInternal()
(f pid=1393046)     @     0x7fdf19cd07c4        128  ray::core::ReferenceCounter::GetAndClearLocalBorrowersInternal()
(f pid=1393046)     @     0x7fdf19cd07c4        128  ray::core::ReferenceCounter::GetAndClearLocalBorrowersInternal()
(f pid=1393046)     @ ... and at least 1000 more frames
(f pid=1393046) [2022-03-02 11:59:26,565 E 1393046 1393046] logging.cc:321: *** SIGSEGV received at time=1646251166 on cpu 7 ***
(f pid=1393046) [2022-03-02 11:59:26,565 E 1393046 1393046] logging.cc:321: PC: @     0x7fdf19cc9c88  (unknown)  absl::lts_20211102::container_internal::raw_hash_set<>::find<>()
(f pid=1393046) [2022-03-02 11:59:26,566 E 1393046 1393046] logging.cc:321:     @     0x7fdf1b3f2210  2140023952  (unknown)
(f pid=1393046) [2022-03-02 11:59:26,566 E 1393046 1393046] logging.cc:321:     @     0x7fdf19cd05bd        128  ray::core::ReferenceCounter::GetAndClearLocalBorrowersInternal()
(f pid=1393046) [2022-03-02 11:59:26,566 E 1393046 1393046] logging.cc:321:     @     0x7fdf19cd07c4        128  ray::core::ReferenceCounter::GetAndClearLocalBorrowersInternal()
(f pid=1393046) [2022-03-02 11:59:26,566 E 1393046 1393046] logging.cc:321:     @     0x7fdf19cd07c4        128  ray::core::ReferenceCounter::GetAndClearLocalBorrowersInternal()
(f pid=1393046) [2022-03-02 11:59:26,566 E 1393046 1393046] logging.cc:321:     @     0x7fdf19cd07c4        128  ray::core::ReferenceCounter::GetAndClearLocalBorrowersInternal()
(f pid=1393046) [2022-03-02 11:59:26,566 E 1393046 1393046] logging.cc:321:     @     0x7fdf19cd07c4        128  ray::core::ReferenceCounter::GetAndClearLocalBorrowersInternal()
(f pid=1393046) [2022-03-02 11:59:26,566 E 1393046 1393046] logging.cc:321:     @     0x7fdf19cd07c4        128  ray::core::ReferenceCounter::GetAndClearLocalBorrowersInternal()
(f pid=1393046) [2022-03-02 11:59:26,566 E 1393046 1393046] logging.cc:321:     @     0x7fdf19cd07c4        128  ray::core::ReferenceCounter::GetAndClearLocalBorrowersInternal()
(f pid=1393046) [2022-03-02 11:59:26,566 E 1393046 1393046] logging.cc:321:     @     0x7fdf19cd07c4        128  ray::core::ReferenceCounter::GetAndClearLocalBorrowersInternal()
(f pid=1393046) [2022-03-02 11:59:26,566 E 1393046 1393046] logging.cc:321:     @     0x7fdf19cd07c4        128  ray::core::ReferenceCounter::GetAndClearLocalBorrowersInternal()
(f pid=1393046) [2022-03-02 11:59:26,566 E 1393046 1393046] logging.cc:321:     @     0x7fdf19cd07c4        128  ray::core::ReferenceCounter::GetAndClearLocalBorrowersInternal()
(f pid=1393046) [2022-03-02 11:59:26,566 E 1393046 1393046] logging.cc:321:     @     0x7fdf19cd07c4        128  ray::core::ReferenceCounter::GetAndClearLocalBorrowersInternal()
(f pid=1393046) [2022-03-02 11:59:26,566 E 1393046 1393046] logging.cc:321:     @     0x7fdf19cd07c4        128  ray::core::ReferenceCounter::GetAndClearLocalBorrowersInternal()
(f pid=1393046) [2022-03-02 11:59:26,566 E 1393046 1393046] logging.cc:321:     @     0x7fdf19cd07c4        128  ray::core::ReferenceCounter::GetAndClearLocalBorrowersInternal()
(f pid=1393046) [2022-03-02 11:59:26,566 E 1393046 1393046] logging.cc:321:     @     0x7fdf19cd07c4        128  ray::core::ReferenceCounter::GetAndClearLocalBorrowersInternal()
(f pid=1393046) [2022-03-02 11:59:26,566 E 1393046 1393046] logging.cc:321:     @     0x7fdf19cd07c4        128  ray::core::ReferenceCounter::GetAndClearLocalBorrowersInternal()
(f pid=1393046) [2022-03-02 11:59:26,566 E 1393046 1393046] logging.cc:321:     @     0x7fdf19cd07c4        128  ray::core::ReferenceCounter::GetAndClearLocalBorrowersInternal()
(f pid=1393046) [2022-03-02 11:59:26,566 E 1393046 1393046] logging.cc:321:     @     0x7fdf19cd07c4        128  ray::core::ReferenceCounter::GetAndClearLocalBorrowersInternal()
(f pid=1393046) [2022-03-02 11:59:26,566 E 1393046 1393046] logging.cc:321:     @     0x7fdf19cd07c4        128  ray::core::ReferenceCounter::GetAndClearLocalBorrowersInternal()
(f pid=1393046) [2022-03-02 11:59:26,566 E 1393046 1393046] logging.cc:321:     @     0x7fdf19cd07c4        128  ray::core::ReferenceCounter::GetAndClearLocalBorrowersInternal()
(f pid=1393046) [2022-03-02 11:59:26,566 E 1393046 1393046] logging.cc:321:     @     0x7fdf19cd07c4        128  ray::core::ReferenceCounter::GetAndClearLocalBorrowersInternal()
(f pid=1393046) [2022-03-02 11:59:26,566 E 1393046 1393046] logging.cc:321:     @     0x7fdf19cd07c4        128  ray::core::ReferenceCounter::GetAndClearLocalBorrowersInternal()
(f pid=1393046) [2022-03-02 11:59:26,566 E 1393046 1393046] logging.cc:321:     @     0x7fdf19cd07c4        128  ray::core::ReferenceCounter::GetAndClearLocalBorrowersInternal()
(f pid=1393046) [2022-03-02 11:59:26,566 E 1393046 1393046] logging.cc:321:     @     0x7fdf19cd07c4        128  ray::core::ReferenceCounter::GetAndClearLocalBorrowersInternal()
```

cc @stephanie-wang @jjyao

### PR DESCRIPTION
**Pull request by jjyao**
## Why are these changes needed?

When we deserialize actor handle via pickle, we will register it with an outer object ref equaling to itself which is wrong. For out-of-band deserialization, there should be no outer object ref.

## Related issue number
Closes #22780

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/tests/test_reference_counting_2.py [hunk 1]
```diff
@@ -699,6 +700,25 @@ def pass_nested_ref(borrower, outer_ref):
...
+
+    actor = Actor.remote()
+
+    @ray.remote
+    def func(config):
+        # deep copy will pickle and unpickle the actor handle.
+        config = copy.deepcopy(config)
+        return ray.get(config["actor"].ping.remote())
+
+    assert ray.get(func.remote({"actor": actor})) == 1
```
Comment (stephanie-wang) [inline python/ray/tests/test_reference_counting_2.py:719]: Can we also test that the unpickled actor handle will receive an error if we try to use it after the original handles are gone?

File: python/ray/tests/test_reference_counting_2.py [hunk 2]
```diff
@@ -699,6 +706,46 @@ def pass_nested_ref(borrower, outer_ref):
...
+
+
+def test_out_of_band_actor_handle_bypass_reference_counting(shutdown_only):
```
Comment (stephanie-wang) [inline python/ray/tests/test_reference_counting_2.py:728]: Could we only have this test case? Or does the other one add additional value?

**Review by stephanie-wang — APPROVED** [review]

File: python/ray/tests/test_reference_counting_2.py (hunk 2 above)
Comment (jjyao) [reply python/ray/tests/test_reference_counting_2.py:728]: Yea, the other one tests the borrowed case (that triggered the bug that this PR is fixing).

**Comment by jjyao** [comment]
Failed tests are unrelated.

**Pull request merged**